Sources/PointFree/Account/Invite.swift

```diff
@@ -199,8 +267,60 @@ private func validateIsNot(currentUser: Database.User) -> (Database.User) -> Eit
   }
 }
 
-private func validateDoesNotHaveSubscription(user: Database.User) -> EitherIO<Error, Database.User> {
-  return user.subscriptionId != nil
-    ? lift(.left(unit))
-    : lift(.right(user))
+private func validateActiveStripeSubscription(
+  _ errorOrSubscription: Either<Prelude.Unit, Stripe.Subscription>
+  )
+  -> Either<Prelude.Unit, Stripe.Subscription> {
+
+    return errorOrSubscription.flatMap { stripeSubscription in
+      stripeSubscription.status != .active ? .left(unit) : .right(stripeSubscription)
+    }
+}
+
+private func redirectCurrentSubscribers<A, B>(
+  _ middleware: @escaping Middleware<StatusLineOpen, ResponseEnded, T3<A, Database.User?, B>, Data>
+  ) -> Middleware<StatusLineOpen, ResponseEnded, T3<A, Database.User?, B>, Data> {
+
+  return { conn in
+    guard
+      let user = get2(conn.data),
+      let subscriptionId = user.subscriptionId
+      else { return middleware(conn) }
+
+    let hasActiveSubscription = AppEnvironment.current.database.fetchSubscriptionById(subscriptionId)
+      .mapExcept(requireSome)
+      .bimap(const(unit), id)
+      .flatMap { AppEnvironment.current.stripe.fetchSubscription($0.stripeSubscriptionId) }
+      .run
+      .map { $0.right?.status == .some(.active) }
+
+    return hasActiveSubscription.flatMap {
+      $0
+        ? (conn |> redirect(to: .account(.index), headersMiddleware: flash(.warning, "You already have an active subscription. If you want to accept this team invite you must cancel your current subscription.")))
+        : middleware(conn)
+    }
+  }
+}
+
+private func validateCurrentUserIsNotInviter<A>(_ data: T3<Database.TeamInvite, Database.User?, A>) -> Bool {
+  let (teamInvite, currentUser) = (get1(data), get2(data))
+  return currentUser?.id.unwrap != .some(teamInvite.inviterUserId.unwrap)
+}
+
+private func validateCurrentUserIsInviter<A>(_ data: T3<Database.TeamInvite, Database.User, A>) -> Bool {
+  let (teamInvite, currentUser) = (get1(data), get2(data))
+  return currentUser.id.unwrap == teamInvite.inviterUserId.unwrap
+}
+
+private func validateEmailDoesNotBelongToInviter<A>(_ data: T3<EmailAddress, Database.User, A>) -> Bool {
+  let (email, inviter) = (get1(data), get2(data))
+  return email.unwrap.lowercased() != inviter.email.unwrap.lowercased()
+}
+
+private func fetchTeamInviter<A>(_ data: T2<Database.TeamInvite, A>) -> IO<T3<Database.TeamInvite, Database.User, A>?> {
```

Comment:
I'm making use of the `(A) -> IO<B?>` variant of `filterMap`! that was a good idea!